### PR TITLE
ignore: introduce .dvcignore

### LINKED ISSUE #1499: dvc: support .dvcignore
Same as .gitignore, but for dvc. It will tell dvc which paths to ignore when caching data. Would be extremely useful for ignoring auto-generated/temporary files/directories that are created as a side effect and don't carry anything useful.

First iteration should make patterns listed in dvcignore be ignored on Repo.stages() when collecting stages.
```
echo '/directory_with_millions_of_files' > .dvcignore
dvc status # should not enter `directory_with_millons_of_files`
```

Second iteration should support `dvc add/run`. We have a separate issue for it at https://github.com/iterative/dvc/issues/1876

### PR DESCRIPTION
**Pull request by pared**
Fixes #1499 
Related to #1863 
Note: This implementation allows to ignore files via `.dvcignore` file placed in same directory as ignored file, or any parent directory. This behaviour is consistent with [`.gitignore` ignore rules](https://git-scm.com/docs/gitignore).

### CODE REVIEW

File: dvc/ignore.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    def __init__(self):
+        self.ignores = [
+            DvcIgnoreDir(".git"),
+            DvcIgnoreDir(".hg"),
+            DvcIgnoreDir(".dvc"),
+        ]
+
+    def __call__(self, *args, **kwargs):
+        assert len(args) == 3
+        root, dirs, files = args
```
Comment (efiop) [inline dvc/ignore.py]: why not just use these three in the __call__ itself?

File: dvc/ignore.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        for pattern in self.patterns:
+            if fnmatch(full_path, pattern) or fnmatch(rel_path, pattern):
+                return True
+
+        return False
+
+    def _read_lines_ignore_comments(self, file_path):
+        with open(file_path, "r") as fobj:
+            return [l for l in fobj if not l.lstrip().startswith("#")]
```
Comment (efiop) [inline dvc/ignore.py]: If we don't support `!` syntax, it is a good idea to ignore such entries and/or put a warning and/or even error about it.

File: dvc/scm/tree.py [hunk 3]
```diff
@@ -57,9 +58,7 @@ def walk(self, top, topdown=True):
         def onerror(e):
             raise e
 
-        for root, dirs, files in os.walk(
+        for root, dirs, files in dvc_walk(
```
Comment (efiop) [inline dvc/scm/tree.py]: Shouldn't we apply dvcignore logic to other trees as well?

Comment (efiop) [reply dvc/scm/tree.py]: Meaning, we should probably ignore dvcignore'd paths when iterating over git tree too.

File: dvc/ignore.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        files = [f for f in files if not self.matches(root, f)]
+        dirs = [d for d in dirs if not self.matches(root, d)]
+
+        return dirs, files
+
+    def matches(self, root, basename):
+        full_path = os.path.join(root, basename)
+        rel_path = os.path.relpath(full_path, self.dirname)
+
+        for pattern in self.patterns:
```
Comment (efiop) [inline dvc/ignore.py]: Patterns that we are using here should be unixpaths, right? But you are using os.path above, so on windows it is going to be ntpath. In your tests, you are only testing on unixpaths, which is probably why it works on windows. Or am I missing something?

Comment (efiop) [reply dvc/ignore.py]: EDIT: you are indeed using os.path in tests, sorry didn't notice right away 🙂Is it fnmatch that handles path transformations?

File: dvc/utils/__init__.py [hunk 5]
```diff
@@ -245,7 +246,24 @@ def load_stage_file_fobj(fobj, path):
         raise StageFileCorruptedError(path)
 
 
+def dvc_walk(top, topdown=True, onerror=None, followlinks=False):
+    """
+    Proxy for `os.walk` directory tree generator.
+    Utilizes DvcIgnoreFilter functionality.
+    """
+    ignore_filter = DvcIgnoreFilter()
```
Comment (efiop) [inline dvc/utils/__init__.py]: Here you are creating a new filter, that will only be able to account for dvcignores that are present in `top` dir and down below, and will ignore dvcignores specified in parent directory.

File: dvc/utils/fs.py [hunk 6]
```diff
@@ -20,7 +21,7 @@ def get_mtime_and_size(path):
     mtime = os.path.getmtime(path)
 
     if os.path.isdir(path):
-        for root, dirs, files in os.walk(str(path)):
+        for root, dirs, files in dvc_walk(str(path)):
```
Comment (efiop) [inline dvc/utils/fs.py]: This will change mtime/size logic for directories. Did you test it with `dvc add directory` that has ignored files in it? What if one of that ignored files changes? It will change mtime of the directory itself, right? Which will make state entry not valid and will make dvc re-compute checksum for it. Should we not account for mtimes and sizes of the directories itself and only rely on files?

File: dvc/ignore.py (hunk 4 above)
Comment (pared) [reply dvc/ignore.py]: Ineed, fnmatch handles patterns:
https://github.com/python/cpython/blob/cb0748d3939c31168ab5d3b80e3677494497d5e3/Lib/fnmatch.py#L22

Comment (efiop) [reply dvc/ignore.py]: @pared Got it! Thanks!

File: dvc/utils/__init__.py (hunk 5 above)
Comment (pared) [reply dvc/utils/__init__.py]: That could be dangerous, thanks for pointing out!

File: dvc/utils/fs.py (hunk 6 above)
Comment (pared) [reply dvc/utils/fs.py]: Summary of private conversation about get_mtime_and_size:
We have a problem when we want to use get_mtime_and_size when it contains .dvcignore-ed paths because:
- we should not take into account ignored  files, because updating ignored file will result in updated directory mtime, which will (for example) cause checksum recalculation, when actual data (not ignored) has not been modified


Straightforward solution for that is to analyse files only, however that causes another problem:
- renaming file under directory will not cause update of this file mtime, but it will cause update of parent directory mtime. So, in such case, analysing files only will not return changed mtime, which is obviously not true.

We need some smarter way to handle get_mtime_and_size.

Comment (pared) [reply dvc/utils/fs.py]: Summary of another private conversation:
- Because of resons mentioned in post above, we need mechanism that will detect changes in dir, but also will be immune to changes performed on "ignored" files.

New way of dealing with this:
1. Let size be calculated as is, because modifications of ignored files will not affect dir size calculations (duirng walk we ignore those files)
2. Develop new method, that will be used to detect changes, effectively replacing mtime update. How to do that? Put all files for given dir in list `[(relpath: mtime)]` and in the end don't extract greatest mtime(as it is done now), but pass this object through checksum method. Thanks to that we will be able to detect changes in directory tree, yet we will not have to worry about update of ignored files updating mtime of parent (and not ignored) directory.

File: requirements.txt [hunk 7]
```diff
@@ -28,3 +28,4 @@ appdirs>=1.4.3
 treelib>=1.5.5
 inflect>=2.1.0
 humanize>=0.5.1
+dulwich>=0.19.11
```
Comment (efiop) [inline requirements.txt]: Need to add it to setup.py too.

File: dvc/utils/fs.py [hunk 8]
```diff
@@ -15,23 +16,33 @@ def get_inode(path):
...
+
+            for file in files:
+                entry = os.path.join(root, file)
                 stat = os.stat(entry)
                 size += stat.st_size
-                entry_mtime = stat.st_mtime
-                if entry_mtime > mtime:
-                    mtime = entry_mtime
+                files_mtimes[entry] = stat.st_mtime
+        # we need to take into account that analysed dir can have .dvcignored
```
Comment (efiop) [inline dvc/utils/fs.py]: Let's add a better explanation why we do this. Also, `files_mtimes` dict is not sorted, you need to use a list and actually sort it by relpath. See https://github.com/iterative/dvc/blob/master/dvc/remote/local.py#L251 . Otherwise the result of dict_md5 might not be reproducible. Let's add a test that is making sure that mtime_and_size is the same for the same dir.

File: dvc/ignore.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
+from dulwich.ignore import match_pattern, read_ignore_patterns
+from dvc.utils.compat import cast_bytes
+
+import os
```
Comment (efiop) [inline dvc/ignore.py]: standard imports should be higher than others :) Pylint would catch that, need to push a little and finally enable it :slightly_smiling_face:

File: dvc/utils/fs.py (hunk 8 above)
Comment (efiop) [reply dvc/utils/fs.py]: @pared pointed out that I'm wrong about this. Indeed, dict representation _is_ sorted in dict_md5 itself, so we should be fine. The reason I've used lists and sorted them in dir cache is that there I have opted for storing {"relpath": relpath, "md5": md5} dicts just in case we would need to add something else to them later. It has been a long time and we didn't add anything there yet, so that might've been for nothing :slightly_smiling_face:

**Comment by efiop** [comment]
@ei-grad Please take a look. :slightly_smiling_face:

File: dvc/ignore.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    def _no_negate_pattern_matches(self, path):
+        return all([not match_pattern(path, p) for p in self.negate_patterns])
+
+    def __hash__(self):
+        return hash(self.ignore_file_path)
+
+
+class DvcIgnoreDir(DvcIgnore):
+    def __init__(self, basename):
+        self.basename = basename
```
Comment (efiop) [inline dvc/ignore.py]: CC complains about __init__ being the same as in DvcIgnoreFile. Maybe it makes sense to move it to DvcIgnore.__init__ ?

File: dvc/ignore.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+
+class DvcIgnore(object):
+    DVCIGNORE_FILE = ".dvcignore"
+
+    def __call__(self, root, dirs, files):
+        raise NotImplementedError
+
+
+class DvcIgnoreFromFile(DvcIgnore):
```
Comment (efiop) [inline dvc/ignore.py]: Is there a point in inheriting from DvcIgnore ?

File: dvc/scm/tree.py [hunk 12]
```diff
@@ -57,9 +66,13 @@ def walk(self, top, topdown=True):
...
+            topdown=topdown,
+            onerror=onerror,
+            ignore_file_handler=ignore_file_handler,
         ):
-            if topdown:
-                dirs[:] = [i for i in dirs if i not in (".git", ".hg", ".dvc")]
             yield os.path.normpath(root), dirs, files
+
+    def get_tree_root(self):
+        return self.repo_root
```
Comment (efiop) [inline dvc/scm/tree.py]: but `repo_root` is already public, why add a getter on top?

File: dvc/ignore.py (hunk 11 above)
Comment (pared) [reply dvc/ignore.py]: Just inheriting __call__, I thought all ignores should have this

File: dvc/utils/fs.py [hunk 13]
```diff
@@ -63,3 +64,18 @@ def contains_symlink_up_to(path, base_path):
     if os.path.dirname(path) == path:
         return False
     return contains_symlink_up_to(os.path.dirname(path), base_path)
+
+
+def get_parent_dirs_up_to(dir, root_dir):
+    dir = os.path.normpath(dir)
```
Comment (efiop) [inline dvc/utils/fs.py]: Please don't use `dir` as a var name, since it overlaps with the standard `dir()` function.

File: dvc/ignore.py (hunk 10 above)
Comment (pared) [reply dvc/ignore.py]: DvcIgnoreFromFile requires different initialization, but we can extract some intermediate object

File: dvc/utils/fs.py [hunk 14]
```diff
@@ -63,3 +64,18 @@ def contains_symlink_up_to(path, base_path):
...
+    root_dir = os.path.normpath(root_dir)
+    if root_dir not in dir:
+        return []
+
+    dirs = []
+    dirs.append(dir)
+    while dir != root_dir:
+        dir, basename = os.path.split(dir)
```
Comment (efiop) [inline dvc/utils/fs.py]: if you only need parent, you could use `os.path.dirname(dir)`

File: dvc/scm/tree.py (hunk 12 above)
Comment (pared) [reply dvc/scm/tree.py]: To unify method of accessing Tree root, this method looks different for example in GitTree. However we are not using it there so maybe there is no point in doing that yet.

File: dvc/utils/fs.py [hunk 15]
```diff
@@ -63,3 +64,18 @@ def contains_symlink_up_to(path, base_path):
     if os.path.dirname(path) == path:
         return False
     return contains_symlink_up_to(os.path.dirname(path), base_path)
+
+
+def get_parent_dirs_up_to(dir, root_dir):
+    dir = os.path.normpath(dir)
+    root_dir = os.path.normpath(root_dir)
```
Comment (efiop) [inline dvc/utils/fs.py]: you would probably need to make sure that root_dir and dir are absolute paths, right? Btw, please add a unit test for it.

File: tests/test_stage.py [hunk 16]
```diff
@@ -165,3 +165,27 @@ def test_remote_dependency(self):
         assert main(["import", "remote://storage/file", "movie.txt"]) == 0
 
         assert os.path.exists("movie.txt")
+
+
+class TestIgnore(TestDvc):
```
Comment (efiop) [inline tests/test_stage.py]: This test is not related to dvc/stages.py, but rather to dvc/repo, so it should be in `tests/test_repo.py`.

File: dvc/utils/fs.py (hunk 14 above)
Comment (pared) [reply dvc/utils/fs.py]: sorry, some leftover after refactors

File: dvc/ignore.py (hunk 10 above)
Comment (efiop) [reply dvc/ignore.py]: Got it :)

File: dvc/ignore.py (hunk 11 above)
Comment (efiop) [reply dvc/ignore.py]: Ah, sorry, didn't catch that.

File: dvc/scm/tree.py (hunk 12 above)
Comment (efiop) [reply dvc/scm/tree.py]: I see. So maybe just wrap it into a property?

File: tests/test_ignore.py [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        self.ignore_file_handler = DvcIgnoreFileHandler(self.dvc.tree)
+
+    def _get_all_paths(self):
+
+        paths = []
+        ignore_file_handler = DvcIgnoreFileHandler(self.dvc.tree)
+        for root, dirs, files in self.dvc.tree.walk(
+            self.dvc.root_dir, ignore_file_handler=ignore_file_handler
+        ):
+            for dir in dirs:
```
Comment (efiop) [inline tests/test_ignore.py]: Please don't use `dir` and `file` as var names, because they highlight as builtins.
